### PR TITLE
refactor: switch to new API for recording deps

### PR DESCRIPTION
**Pull request by rgrinberg**
Continuation of the splitting of https://github.com/ocaml/dune/pull/9132

### CODE REVIEW

**Review by snowleopard — APPROVED** [review]
Looks good though isn't this all that's left in the original PR? We might as well merge it.

**Comment by rgrinberg** [comment]
There's one last PR i'm working on which makes `eval_mode` private. This requires changing the implementation of a few helpers in `dune_rules/action_builder.ml`

**Comment by snowleopard** [comment]
Cool, almost done :)

File: src/dune_rules/action_builder.ml [hunk 1]
```diff
@@ -42,13 +42,8 @@ let dyn_paths paths = dyn_deps (paths >>| fun (x, paths) -> x, Dep.Set.of_files
...
 let contents p =
-  of_thunk
-    { f =
-        (fun _mode ->
-          let open Memo.O in
-          let+ x = Build_system.read_file p in
-          x, Dep.Map.empty)
-    }
+  let* () = Dep.file p |> Dep.Set.singleton |> Build_system.record_deps in
+  of_memo (Build_system.read_file p)
```
Comment (Alizter) [inline src/dune_rules/action_builder.ml:46]: @rgrinberg This wasn't happening before and Coq rules were relying on it. Removing the recording of the dep fixes #10149.

Comment (snowleopard) [reply src/dune_rules/action_builder.ml:46]: Huh, indeed! Good catch.

**Pull request merged**